tensorflow_serving/batching/batching_util.cc

```diff
@@ -0,0 +1,229 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow_serving/batching/batching_util.h"
+
+#include <string>
+
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/lib/core/errors.h"
+
+
+namespace tensorflow {
+namespace serving {
+
+// Padding for one dimension of a tensor.
+struct OneDimPadding {
+  int64 first;  // pad before
+  int64 second;  // pad after
+};
+
+// Constructs array of paddings, where:
+// paddings[i].first - number of objects to add before elements in dimension i
+// of given tensor,
+// paddings[i].second - number of objects to add after elements in dimension i.
+// This padding signature is used when performing internal padding with Eigen.
+//
+// When building paddings it is assumed that tensor needs to be padded
+// after each dimension, so its shape matches max_dim_sizes,
+// First entry of max_dim_sizes, which is maximum size of zeroth dimension,
+// is ignored, because we don't perform padding in that dimension.
+template<int num_dims>
+std::array<OneDimPadding, num_dims> CreatePadding(Tensor tensor,
+        const std::vector<int>& max_dim_sizes) {
+  std::array<OneDimPadding, num_dims> padding;
+  for (unsigned int i = 0; i < max_dim_sizes.size(); ++i) {
+    if (i > 0 && max_dim_sizes[i] - tensor.dim_size(i) > 0) {
+      padding[i] = {0, max_dim_sizes[i] - tensor.dim_size(i)};
+    } else {
+      padding[i] = {0, 0};
+    }
+  }
+  return padding;
+}
+
+// Functor, which performs padding of given input tensor
+// using specified padding signature.
+// For example, given tensor of shape [1, 2, 3] and padding signature
+// [[0, 0], [0, 2], [2, 2]]
+// functor produces padded_tensor of shape [1, 4, 7].
+template <typename T, int num_dims>
+struct PadTensor {
+  Status operator()(Tensor input,
+      const std::array<OneDimPadding, num_dims>& padding,
+      Tensor* output) {
+    TensorShape output_shape;
+    for (int d = 0; d < num_dims; ++d) {
+      // Pad before existing elements.
+      const int32 before_d = padding[d].first;
+      // Pad after existing elements.
+      const int32 after_d = padding[d].second;
+      output_shape.AddDim(before_d + input.dim_size(d) + after_d);
+    }
+    if (output_shape.num_elements() == input.NumElements()) {
+      auto result = output->CopyFrom(input, output_shape);
```

Comment:
Please avoid "auto" except for iterators (which have gnarly types that aren't usually helpful for readability of the code). Please find all non-iterator uses of "auto" in your change and put in the real time. Thanks!